aten/src/ATen/function_wrapper.py

```diff
@@ -637,7 +732,10 @@ def device_guard(option, dispatch_options, dispatch_tensor):
     # For factory methods the `DeviceGuard` is already in the template.
     if option.get('device_guard', True):
         if dispatch_options:
-            return 'const DeviceGuard device_guard({}.device());'.format(dispatch_options['name'])
+            if any(arg['type'] == 'c10::optional<ScalarType>' for arg in option['arguments']):
```

Comment:
Why is this test on `ScalarType`? A test on `DeviceType` would be more logical...